distributed/client.py

```diff
@@ -90,6 +91,7 @@
 _global_clients = weakref.WeakValueDictionary()
 _global_client_index = [0]
 
+_deserialize_client = ContextVar("_deserialize_client", default=None)
```

Comment:
Do we need to use contextvars here? We currently use a thread local for storing these kind of things, but that doesn't always play nicely with async. Is there a way we could refactor part of distributed to avoid passing around this implicit state?

In general I think`contextvars` can be quite useful where required, but want to ensure it's needed here before bringing it in. We also already have a fair bit of implicit state, I'm a bit hesitant to add more if we can avoid it.

Reply:
To avoid using contextvars you need to explicitly propagate the client until you stop calling async functions and just call sync functions. In this case, this means that

- ``client.Client._get_dataset`` needs to pass ``self`` to
- ``core.send_recv``, which in turn needs to pass it to
- ``comm.tcp.TCP.read`` (plus its variants ``comm.inproc.InProc.read`` and the 100% untested ``comm.ucx.UCX.read``), which in turn needs to pass it to
- ``comm.utils.from_frames``, which can finally put a context manager (setting a thread-local variable) around
https://github.com/dask/distributed/blob/8376f227e288757c3fd9b1e8742ca350e6e57b25/distributed/comm/utils.py#L60-L62
- which will alter the behaviour of ``client.Future.__getstate__``

and it should be really self-evident why I don't recommend doing it.

> We currently use a thread local for storing these kind of things

For this one you were using a global, ``client._global_clients``.
contextvars exactly fits the use case of thread locals when you use coroutines.

Reply:
Makes sense to me. Should this logic replace how we access implicit clients everywhere? If so, perhaps we should rename the context var key?

Reply:
Are there other cases where a method of Client loses ``self`` along the way and re-acquires it later from a global variable?

Also note the very big fat caveat of the race conditions on Python 3.6. Not sure how much of the dask.distributed user base still uses Python 3.6 AND has more than one client running.

Reply:
Sorry, let me clarify.

It feels a bit weird to handle setting the contextvar in the `_get_dataset` method. It might be cleaner to have a parent context over all coroutines started by a client that includes a reference back to the client. As it stands this is baking in functionality to `Future` that is only used by one method, when we likely want this everywhere.

Reply:
> It might be cleaner to have a parent context over all coroutines started by a client that includes a reference back to the client. 

Are you talking to some sort of blanket decorator applied around all methods of Client? If so, it feels a bit overkill to me, as I do not know of another method that needs it.

>  As it stands this is baking in functionality to Future that is only used by one method, when we likely want this everywhere.

You definitely want to reuse this same variable every time you have a ``self`` reference to a Client instance, invoke a method, lose the reference (because it would be too cumbersome to propagated by hand), and then you need to reacquire it deeper down the line _before the end of the method_. If you can point me to any other function that does this, I'll be happy to look into it.

Reply:
@jcrist ping

Reply:
I'm redesigning the change to make it more generic, hold on